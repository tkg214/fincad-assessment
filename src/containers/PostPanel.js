import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Panel, Accordion, Col } from 'react-bootstrap';
import FontAwesome from 'react-fontawesome';

import { fetchComments } from '../actions/comments';
import CommentList from '../components/CommentList';
import Post from '../components/Post';
import NewComment from './NewComment';

class PostPanel extends Component {
  constructor(props) {
    super(props);
    this.state = {
      commentView: 'View Comments',
      commentsVisible: false,
      bodyInput: '',
      commentIcon: 'caret-square-o-right',
      activeKey: ''
    };
    this.handleToggle = this.handleToggle.bind(this);
    this.handleCommentSubmit = this.handleCommentSubmit.bind(this);
  }

  handleToggle(activeKey) {
    !this.state.commentsVisible &&
      this.props.comments.length === 0 &&
      this.props.fetchComments(this.props.id);
    this.state.commentsVisible
      ? this.setState({
          commentView: 'View Comments',
          commentsVisible: false,
          commentIcon: 'caret-square-o-right',
          activeKey: ''
        })
      : this.setState({
          commentView: 'Close Comments',
          commentsVisible: true,
          commentIcon: 'caret-square-o-down',
          activeKey
        });
  }

  handleCommentSubmit() {
    this.setState({
      commentView: 'Close Comments',
      commentsVisible: true,
      commentIcon: 'caret-square-o-down',
      activeKey: '1'
    });
  }

  render() {
    const { title, body, name, comments, id } = this.props;
    const postHeader = (
      <h2>
        {title}
      </h2>
    );

    const commentsHeader = (
      <p>
        <FontAwesome name={this.state.commentIcon} />
        &nbsp; {this.state.commentView}
      </p>
    );
    const commentCount = comments.length;

    return (
      <div>
        <Panel header={postHeader}>
          <Post data-test="post" name={name} body={body} />
          <Accordion
            activeKey={this.state.activeKey}
            onSelect={this.handleToggle}
          >
            <Panel header={commentsHeader} eventKey="1">
              <Col md={10} mdOffset={2}>
                {commentCount > 0
                  ? <CommentList
                      comments={comments}
                      commentCount={commentCount}
                    />
                  : <p>{`There are no comments for this post.`}</p>}
              </Col>
            </Panel>
          </Accordion>
          {this.props.auth.isLoggedIn &&
            <NewComment
              commentsVisible={this.state.commentsVisible}
              handleCommentSubmit={this.handleCommentSubmit}
              commentCount={commentCount}
              postId={id}
              data-test="newComment"
            />}
        </Panel>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  auth: state.auth
});

const mapDispatchToProps = dispatch => ({
  fetchComments: postId => {
    dispatch(fetchComments(postId));
  }
});

export { PostPanel as Component };
export default connect(mapStateToProps, mapDispatchToProps)(PostPanel);
