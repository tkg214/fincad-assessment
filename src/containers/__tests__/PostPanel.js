import React from 'react';
import { shallow } from 'enzyme';

import { Component as PostPanel } from '../PostPanel';

test('the post panel renders correctly', () => {
  const wrapper = shallowRenderPost();
  expect(wrapper).toMatchSnapshot();
});

test('the new comment component does not render if user is not logged in', () => {
  const wrapper = shallowRenderPost();
  expect(wrapper.find(sel('newComment')).node).toBeUndefined;
});

test('the new comment component renders if user is logged in', () => {
  const auth = { isLoggedIn: true };
  const wrapper = shallowRenderPost({ auth });
  expect(wrapper.find(sel('newComment')).node).toBeDefined;
});

function getPostProps(props) {
  return {
    auth: { isLoggedIn: false },
    title: '',
    name: '',
    body: '',
    comments: [],
    ...props
  };
}

function shallowRenderPost(props = {}) {
  const postProps = getPostProps(props);
  return shallow(<PostPanel {...postProps} />);
}

function sel(id) {
  return `[data-test="${id}"]`;
}
