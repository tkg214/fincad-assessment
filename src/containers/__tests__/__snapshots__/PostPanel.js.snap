// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the post panel renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <Panel
        bsClass="panel"
        bsStyle="default"
        defaultExpanded={false}
        header={<h2 />}
    >
        <Post
            body=""
            data-test="post"
            name=""
        />
        <Accordion
            activeKey=""
            onSelect={[Function]}
        >
            <Panel
                bsClass="panel"
                bsStyle="default"
                defaultExpanded={false}
                eventKey="1"
                header={
                    <p>
                        <FontAwesome
                            name="caret-square-o-right"
                        />
                          
                        View Comments
                    </p>
                }
            >
                <Col
                    bsClass="col"
                    componentClass="div"
                    md={10}
                    mdOffset={2}
                >
                    <p>
                        There are no comments for this post.
                    </p>
                </Col>
            </Panel>
        </Accordion>
    </Panel>
</div>,
  "nodes": Array [
    <div>
      <Panel
            bsClass="panel"
            bsStyle="default"
            defaultExpanded={false}
            header={<h2 />}
      >
            <Post
                  body=""
                  data-test="post"
                  name=""
            />
            <Accordion
                  activeKey=""
                  onSelect={[Function]}
            >
                  <Panel
                        bsClass="panel"
                        bsStyle="default"
                        defaultExpanded={false}
                        eventKey="1"
                        header={
                              <p>
                                    <FontAwesome
                                          name="caret-square-o-right"
                                    />
                                      
                                    View Comments
                              </p>
                        }
                  >
                        <Col
                              bsClass="col"
                              componentClass="div"
                              md={10}
                              mdOffset={2}
                        >
                              <p>
                                    There are no comments for this post.
                              </p>
                        </Col>
                  </Panel>
            </Accordion>
      </Panel>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <PostPanel
        auth={
                Object {
                        "isLoggedIn": false,
                      }
        }
        body=""
        comments={Array []}
        name=""
        title=""
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": PostPanel {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "handleCommentSubmit": [Function],
        "handleToggle": [Function],
        "props": Object {
          "auth": Object {
            "isLoggedIn": false,
          },
          "body": "",
          "comments": Array [],
          "name": "",
          "title": "",
        },
        "refs": Object {},
        "state": Object {
          "activeKey": "",
          "bodyInput": "",
          "commentIcon": "caret-square-o-right",
          "commentView": "View Comments",
          "commentsVisible": false,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <Panel
                    bsClass="panel"
                    bsStyle="default"
                    defaultExpanded={false}
                    header={<h2 />}
          >
                    <Post
                              body=""
                              data-test="post"
                              name=""
                    />
                    <Accordion
                              activeKey=""
                              onSelect={[Function]}
                    >
                              <Panel
                                        bsClass="panel"
                                        bsStyle="default"
                                        defaultExpanded={false}
                                        eventKey="1"
                                        header={
                                                  <p>
                                                            <FontAwesome
                                                                      name="caret-square-o-right"
                                                            />
                                                              
                                                            View Comments
                                                  </p>
                                        }
                              >
                                        <Col
                                                  bsClass="col"
                                                  componentClass="div"
                                                  md={10}
                                                  mdOffset={2}
                                        >
                                                  <p>
                                                            There are no comments for this post.
                                                  </p>
                                        </Col>
                              </Panel>
                    </Accordion>
          </Panel>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <Panel
                    bsClass="panel"
                    bsStyle="default"
                    defaultExpanded={false}
                    header={<h2 />}
          >
                    <Post
                              body=""
                              data-test="post"
                              name=""
                    />
                    <Accordion
                              activeKey=""
                              onSelect={[Function]}
                    >
                              <Panel
                                        bsClass="panel"
                                        bsStyle="default"
                                        defaultExpanded={false}
                                        eventKey="1"
                                        header={
                                                  <p>
                                                            <FontAwesome
                                                                      name="caret-square-o-right"
                                                            />
                                                              
                                                            View Comments
                                                  </p>
                                        }
                              >
                                        <Col
                                                  bsClass="col"
                                                  componentClass="div"
                                                  md={10}
                                                  mdOffset={2}
                                        >
                                                  <p>
                                                            There are no comments for this post.
                                                  </p>
                                        </Col>
                              </Panel>
                    </Accordion>
          </Panel>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <PostPanel
    auth={
        Object {
            "isLoggedIn": false,
          }
    }
    body=""
    comments={Array []}
    name=""
    title=""
/>,
}
`;
